{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "compressed-wrong",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:39.529998Z",
     "start_time": "2021-03-05T08:49:38.413251Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "indie-federation",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-04T12:31:17.586363Z",
     "start_time": "2021-03-04T12:31:17.515266Z"
    }
   },
   "source": [
    "# Variants Output File Tour\n",
    "\n",
    "This notebook gives a short overview on how to use the `variants_output.tsv` file.\n",
    "\n",
    "It represents a simplified version of the full pipeline output (`output/release/built_final.tsv`) with the goal to be more easily consumable for an analyst or a downstream application. That is,\n",
    " * fields are renamed for better consistency and understandability\n",
    " * redunant or too specific fields are removed\n",
    "\n",
    "\n",
    "## Getting Started\n",
    "To get started, download a recent data release tar ball from https://brcaexchange.org/releases and execute the command below to unpack to content to `/tmp`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "missing-switch",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:39.543022Z",
     "start_time": "2021-03-05T08:49:39.532864Z"
    }
   },
   "outputs": [],
   "source": [
    "#!tar zxf [path to tar.gz file] -C /tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "great-prompt",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:39.556973Z",
     "start_time": "2021-03-05T08:49:39.550418Z"
    }
   },
   "outputs": [],
   "source": [
    "tar_ball_out = Path('/tmp/output')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "contrary-tablet",
   "metadata": {},
   "source": [
    "## First Steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "organized-ireland",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:39.572169Z",
     "start_time": "2021-03-05T08:49:39.564342Z"
    }
   },
   "outputs": [],
   "source": [
    "variants_out_path = tar_ball_out/'variants_output.tsv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "allied-blind",
   "metadata": {},
   "source": [
    "The `variants_output.tsv` is a tab separated text file. Missing/NA values are represented using '-'.\n",
    "\n",
    "One row represents one variant. Some key fields appear as first columns, the remaining fields are ordered by category (see below in section [Field Metadata](#category_desc)) and field name.\n",
    "\n",
    "In python, the file be read easily using `pandas`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "visible-boards",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:39.579411Z",
     "start_time": "2021-03-05T08:49:39.574810Z"
    }
   },
   "outputs": [],
   "source": [
    "def read_tsv(path):\n",
    "    return pd.read_csv(path, sep='\\t', na_values='-', low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fitting-advice",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-04T12:36:47.996176Z",
     "start_time": "2021-03-04T12:36:47.989899Z"
    }
   },
   "source": [
    "In R, the file can be read using\n",
    "\n",
    "```\n",
    "df <- read.csv(file = path, sep = '\\t', na.strings=\"-\", header = TRUE)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "copyrighted-semiconductor",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.633358Z",
     "start_time": "2021-03-05T08:49:39.581903Z"
    }
   },
   "outputs": [],
   "source": [
    "df = read_tsv(variants_out_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "varying-bangladesh",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.699461Z",
     "start_time": "2021-03-05T08:49:41.636241Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>genomic_vcf38</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>genomic_hgvs_38</th>\n",
       "      <th>cdna</th>\n",
       "      <th>protein</th>\n",
       "      <th>source</th>\n",
       "      <th>clinical_significance_enigma</th>\n",
       "      <th>...</th>\n",
       "      <th>individuals_lovd</th>\n",
       "      <th>literature_source_exlovd</th>\n",
       "      <th>method_clinvar</th>\n",
       "      <th>remarks_lovd</th>\n",
       "      <th>source_url</th>\n",
       "      <th>submitter_clinvar</th>\n",
       "      <th>submitters_lovd</th>\n",
       "      <th>variant_frequency_lovd</th>\n",
       "      <th>url_enigma</th>\n",
       "      <th>flags_gnomad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32314514:C&gt;T</td>\n",
       "      <td>32314514</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NC_000013.11:g.32314514C&gt;T</td>\n",
       "      <td>NM_000059.3:c.-1193C&gt;T</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not in 442 controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>1/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32314585:T&gt;C</td>\n",
       "      <td>32314585</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NC_000013.11:g.32314585T&gt;C</td>\n",
       "      <td>NM_000059.3:c.-1122T&gt;C</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not in 442 controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>1/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32314680:C&gt;A</td>\n",
       "      <td>32314680</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NC_000013.11:g.32314680C&gt;A</td>\n",
       "      <td>NM_000059.3:c.-1027C&gt;A</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not in 442 controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>2/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32314690:T&gt;C</td>\n",
       "      <td>32314690</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NC_000013.11:g.32314690T&gt;C</td>\n",
       "      <td>NM_000059.3:c.-1017T&gt;C</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not in 442 controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>1/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32314710:T&gt;TG</td>\n",
       "      <td>32314710</td>\n",
       "      <td>T</td>\n",
       "      <td>TG</td>\n",
       "      <td>NC_000013.11:g.32314711dup</td>\n",
       "      <td>NM_000059.3:c.-996dup</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not in 442 controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>1/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 220 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol          genomic_vcf38       pos ref alt  \\\n",
       "0       BRCA2   chr13:g.32314514:C>T  32314514   C   T   \n",
       "1       BRCA2   chr13:g.32314585:T>C  32314585   T   C   \n",
       "2       BRCA2   chr13:g.32314680:C>A  32314680   C   A   \n",
       "3       BRCA2   chr13:g.32314690:T>C  32314690   T   C   \n",
       "4       BRCA2  chr13:g.32314710:T>TG  32314710   T  TG   \n",
       "\n",
       "              genomic_hgvs_38                    cdna          protein source  \\\n",
       "0  NC_000013.11:g.32314514C>T  NM_000059.3:c.-1193C>T  NP_000050.2:p.?   LOVD   \n",
       "1  NC_000013.11:g.32314585T>C  NM_000059.3:c.-1122T>C  NP_000050.2:p.?   LOVD   \n",
       "2  NC_000013.11:g.32314680C>A  NM_000059.3:c.-1027C>A  NP_000050.2:p.?   LOVD   \n",
       "3  NC_000013.11:g.32314690T>C  NM_000059.3:c.-1017T>C  NP_000050.2:p.?   LOVD   \n",
       "4  NC_000013.11:g.32314711dup   NM_000059.3:c.-996dup  NP_000050.2:p.?   LOVD   \n",
       "\n",
       "  clinical_significance_enigma  ...  individuals_lovd  \\\n",
       "0                          NaN  ...               1.0   \n",
       "1                          NaN  ...               1.0   \n",
       "2                          NaN  ...               1.0   \n",
       "3                          NaN  ...               1.0   \n",
       "4                          NaN  ...               1.0   \n",
       "\n",
       "   literature_source_exlovd  method_clinvar         remarks_lovd  source_url  \\\n",
       "0                       NaN             NaN  not in 442 controls         NaN   \n",
       "1                       NaN             NaN  not in 442 controls         NaN   \n",
       "2                       NaN             NaN  not in 442 controls         NaN   \n",
       "3                       NaN             NaN  not in 442 controls         NaN   \n",
       "4                       NaN             NaN  not in 442 controls         NaN   \n",
       "\n",
       "   submitter_clinvar            submitters_lovd  variant_frequency_lovd  \\\n",
       "0                NaN  Lez J Burke (Brisbane,AU)            1/6603 cases   \n",
       "1                NaN  Lez J Burke (Brisbane,AU)            1/6603 cases   \n",
       "2                NaN  Lez J Burke (Brisbane,AU)            2/6603 cases   \n",
       "3                NaN  Lez J Burke (Brisbane,AU)            1/6603 cases   \n",
       "4                NaN  Lez J Burke (Brisbane,AU)            1/6603 cases   \n",
       "\n",
       "   url_enigma  flags_gnomad  \n",
       "0         NaN           NaN  \n",
       "1         NaN           NaN  \n",
       "2         NaN           NaN  \n",
       "3         NaN           NaN  \n",
       "4         NaN           NaN  \n",
       "\n",
       "[5 rows x 220 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "broadband-festival",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.747640Z",
     "start_time": "2021-03-05T08:49:41.706433Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(40416, 220)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "structural-ranch",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T12:55:23.593258Z",
     "start_time": "2021-03-02T12:55:23.589285Z"
    }
   },
   "source": [
    "## Fields"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "provincial-priority",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T12:56:56.824694Z",
     "start_time": "2021-03-02T12:56:56.821189Z"
    }
   },
   "source": [
    "### Field Metadata\n",
    "\n",
    "More information about the many columns can be found in a separate file `variants_output_field_metadata.tsv`. There every row describes a field in the `variants_output.tsv` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "settled-battlefield",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.767049Z",
     "start_time": "2021-03-05T08:49:41.757894Z"
    }
   },
   "outputs": [],
   "source": [
    "field_metadata_path = tar_ball_out/'variants_output_field_metadata.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "normal-essex",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.823502Z",
     "start_time": "2021-03-05T08:49:41.797847Z"
    }
   },
   "outputs": [],
   "source": [
    "field_metadata = read_tsv(field_metadata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "anonymous-nirvana",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.858903Z",
     "start_time": "2021-03-05T08:49:41.826694Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>field_name</th>\n",
       "      <th>category</th>\n",
       "      <th>description</th>\n",
       "      <th>example_value</th>\n",
       "      <th>useful_resources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>allele_count_exome_afr_gnomad</td>\n",
       "      <td>Allele Frequency</td>\n",
       "      <td>Count of observed alternate alleles (variant) ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>allele_count_exome_amr_gnomad</td>\n",
       "      <td>Allele Frequency</td>\n",
       "      <td>Count of observed alternate alleles (variant) ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>allele_count_exome_asj_gnomad</td>\n",
       "      <td>Allele Frequency</td>\n",
       "      <td>Count of observed alternate alleles (variant) ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>allele_count_exome_eas_gnomad</td>\n",
       "      <td>Allele Frequency</td>\n",
       "      <td>Count of observed alternate alleles (variant) ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>allele_count_exome_eas_jpn_gnomad</td>\n",
       "      <td>Allele Frequency</td>\n",
       "      <td>Count of observed alternate alleles (variant) ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          field_name          category  \\\n",
       "0      allele_count_exome_afr_gnomad  Allele Frequency   \n",
       "1      allele_count_exome_amr_gnomad  Allele Frequency   \n",
       "2      allele_count_exome_asj_gnomad  Allele Frequency   \n",
       "3      allele_count_exome_eas_gnomad  Allele Frequency   \n",
       "4  allele_count_exome_eas_jpn_gnomad  Allele Frequency   \n",
       "\n",
       "                                         description example_value  \\\n",
       "0  Count of observed alternate alleles (variant) ...           NaN   \n",
       "1  Count of observed alternate alleles (variant) ...           NaN   \n",
       "2  Count of observed alternate alleles (variant) ...           NaN   \n",
       "3  Count of observed alternate alleles (variant) ...           NaN   \n",
       "4  Count of observed alternate alleles (variant) ...           NaN   \n",
       "\n",
       "  useful_resources  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dangerous-victorian",
   "metadata": {},
   "source": [
    "<a id='category_desc'></a>\n",
    "The different fields were put into broad categories in order to help you find relevant fields more quickly.\n",
    "\n",
    "Here an overview over the available categories along with the number of available fields in that category "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "parallel-milton",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.886503Z",
     "start_time": "2021-03-05T08:49:41.869005Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Allele Frequency         146\n",
       "Provenance                24\n",
       "Nomenclature              23\n",
       "Evidence                  19\n",
       "Clinical Significance      6\n",
       "Remarks                    1\n",
       "Reference                  1\n",
       "Name: category, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_metadata['category'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pediatric-presence",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-04T13:21:14.032309Z",
     "start_time": "2021-03-04T13:21:14.022070Z"
    }
   },
   "source": [
    "For example, here a list of all fields related to the nomenclature of variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "acoustic-touch",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.921938Z",
     "start_time": "2021-03-05T08:49:41.892447Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>field_name</th>\n",
       "      <th>category</th>\n",
       "      <th>description</th>\n",
       "      <th>example_value</th>\n",
       "      <th>useful_resources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>alt</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>The nucleotide(s) that has changed in this var...</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>bic_nomenclature</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>Variant name according to BIC Nomenclature, wh...</td>\n",
       "      <td>5277A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>ca_id</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>ClinGen Allele Registry canonical ID  CA247478059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>cdna</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>HGVS variant alias which references the nucleo...</td>\n",
       "      <td>NM_007294.3:c.5158A&gt;G</td>\n",
       "      <td>https://brcaexchange.org/help#variant-naming-d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>chr</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>The chromosome on which the variant is found</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>clinvaraccession_enigma</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>Accession corresponding to ENIGMA's submission...</td>\n",
       "      <td>SCV000244390.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>gene_symbol</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>HUGO symbol for the gene containing the variant</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>genomic_hgvs_37</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>HGVS representation of the variant relative to...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>genomic_hgvs_38</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>HGVS representation of the variant relative to...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>genomic_vcf37_source</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>HGVS variant alias which references the nucleo...</td>\n",
       "      <td>chr17:g.41215385:T&gt;C</td>\n",
       "      <td>https://brcaexchange.org/help#variant-naming-d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>genomic_vcf38</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>normalized (right aligned w.r.t cDNA) VCF like...</td>\n",
       "      <td>chr17:g.43063368:T&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>genomic_vcf38_source</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>HGVS variant alias which references the nucleo...</td>\n",
       "      <td>chr17:g.43063368:T&gt;C</td>\n",
       "      <td>https://brcaexchange.org/help#variant-naming-d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>pos</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>The position of the variant, per _____ nomencl...</td>\n",
       "      <td>43063368</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>protein</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>HGVS protein nomenclature</td>\n",
       "      <td>NP_009225.1:p.(Thr1720Ala)</td>\n",
       "      <td>https://brcaexchange.org/help#variant-naming-d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>protein_change</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>A shortened abbreviation of the HGVS Protein a...</td>\n",
       "      <td>T1720A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>ref</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>The nucleotide(s) that is in the reference seq...</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>reference_sequence</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>transcript identifier used to keep track of di...</td>\n",
       "      <td>NM_007294.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>rna_lovd</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>LOVD variant identifier in HGVS RNA nomenclature</td>\n",
       "      <td>r.(?)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>scv_clinvar</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>List of ClinVar accessions for each submitted ...</td>\n",
       "      <td>SCV000244390,SCV000743378,SCV000145365,SCV0007...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>scv_version_clinvar</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>Version numbers of ClinVar accessions.  See SC...</td>\n",
       "      <td>1,1,1,1,2,7,11,4,1,3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>synonyms</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>Alternate names for this variant</td>\n",
       "      <td>NM_007294.3.c.5158A&gt;G,T1720A,5277A&gt;G,NC_000017...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>varconsequences</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>Variant consequences, pulled from Ensembl's Va...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>https://useast.ensembl.org/info/genome/variati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>vr_id</td>\n",
       "      <td>Nomenclature</td>\n",
       "      <td>The digest representing the variant (in GRCh38...</td>\n",
       "      <td>ga4gh:VA.kVvzAwHx3AlCzQ8u6b7J9bLNuxKcSnUr</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  field_name      category  \\\n",
       "144                      alt  Nomenclature   \n",
       "148         bic_nomenclature  Nomenclature   \n",
       "149                    ca_id  Nomenclature   \n",
       "151                     cdna  Nomenclature   \n",
       "152                      chr  Nomenclature   \n",
       "157  clinvaraccession_enigma  Nomenclature   \n",
       "177              gene_symbol  Nomenclature   \n",
       "179          genomic_hgvs_37  Nomenclature   \n",
       "180          genomic_hgvs_38  Nomenclature   \n",
       "181     genomic_vcf37_source  Nomenclature   \n",
       "182            genomic_vcf38  Nomenclature   \n",
       "183     genomic_vcf38_source  Nomenclature   \n",
       "196                      pos  Nomenclature   \n",
       "198                  protein  Nomenclature   \n",
       "199           protein_change  Nomenclature   \n",
       "200                      ref  Nomenclature   \n",
       "201       reference_sequence  Nomenclature   \n",
       "203                 rna_lovd  Nomenclature   \n",
       "204              scv_clinvar  Nomenclature   \n",
       "205      scv_version_clinvar  Nomenclature   \n",
       "212                 synonyms  Nomenclature   \n",
       "214          varconsequences  Nomenclature   \n",
       "219                    vr_id  Nomenclature   \n",
       "\n",
       "                                           description  \\\n",
       "144  The nucleotide(s) that has changed in this var...   \n",
       "148  Variant name according to BIC Nomenclature, wh...   \n",
       "149  ClinGen Allele Registry canonical ID  CA247478059   \n",
       "151  HGVS variant alias which references the nucleo...   \n",
       "152       The chromosome on which the variant is found   \n",
       "157  Accession corresponding to ENIGMA's submission...   \n",
       "177    HUGO symbol for the gene containing the variant   \n",
       "179  HGVS representation of the variant relative to...   \n",
       "180  HGVS representation of the variant relative to...   \n",
       "181  HGVS variant alias which references the nucleo...   \n",
       "182  normalized (right aligned w.r.t cDNA) VCF like...   \n",
       "183  HGVS variant alias which references the nucleo...   \n",
       "196  The position of the variant, per _____ nomencl...   \n",
       "198                          HGVS protein nomenclature   \n",
       "199  A shortened abbreviation of the HGVS Protein a...   \n",
       "200  The nucleotide(s) that is in the reference seq...   \n",
       "201  transcript identifier used to keep track of di...   \n",
       "203   LOVD variant identifier in HGVS RNA nomenclature   \n",
       "204  List of ClinVar accessions for each submitted ...   \n",
       "205  Version numbers of ClinVar accessions.  See SC...   \n",
       "212                   Alternate names for this variant   \n",
       "214  Variant consequences, pulled from Ensembl's Va...   \n",
       "219  The digest representing the variant (in GRCh38...   \n",
       "\n",
       "                                         example_value  \\\n",
       "144                                                  C   \n",
       "148                                            5277A>G   \n",
       "149                                                NaN   \n",
       "151                              NM_007294.3:c.5158A>G   \n",
       "152                                                 17   \n",
       "157                                     SCV000244390.1   \n",
       "177                                              BRCA1   \n",
       "179                                                NaN   \n",
       "180                                                NaN   \n",
       "181                               chr17:g.41215385:T>C   \n",
       "182                               chr17:g.43063368:T>C   \n",
       "183                               chr17:g.43063368:T>C   \n",
       "196                                           43063368   \n",
       "198                         NP_009225.1:p.(Thr1720Ala)   \n",
       "199                                             T1720A   \n",
       "200                                                  T   \n",
       "201                                        NM_007294.3   \n",
       "203                                              r.(?)   \n",
       "204  SCV000244390,SCV000743378,SCV000145365,SCV0007...   \n",
       "205                               1,1,1,1,2,7,11,4,1,3   \n",
       "212  NM_007294.3.c.5158A>G,T1720A,5277A>G,NC_000017...   \n",
       "214                                   missense_variant   \n",
       "219          ga4gh:VA.kVvzAwHx3AlCzQ8u6b7J9bLNuxKcSnUr   \n",
       "\n",
       "                                      useful_resources  \n",
       "144                                                NaN  \n",
       "148                                                NaN  \n",
       "149                                                NaN  \n",
       "151  https://brcaexchange.org/help#variant-naming-d...  \n",
       "152                                                NaN  \n",
       "157                                                NaN  \n",
       "177                                                NaN  \n",
       "179                                                NaN  \n",
       "180                                                NaN  \n",
       "181  https://brcaexchange.org/help#variant-naming-d...  \n",
       "182                                                NaN  \n",
       "183  https://brcaexchange.org/help#variant-naming-d...  \n",
       "196                                                NaN  \n",
       "198  https://brcaexchange.org/help#variant-naming-d...  \n",
       "199                                                NaN  \n",
       "200                                                NaN  \n",
       "201                                                NaN  \n",
       "203                                                NaN  \n",
       "204                                                NaN  \n",
       "205                                                NaN  \n",
       "212                                                NaN  \n",
       "214  https://useast.ensembl.org/info/genome/variati...  \n",
       "219                                                NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_metadata.query('category == \"Nomenclature\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "awful-hartford",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T12:55:36.318444Z",
     "start_time": "2021-03-02T12:55:36.314806Z"
    }
   },
   "source": [
    "## Joining Back with the Full File (More Advanced)\n",
    "\n",
    "\n",
    "In case a field is of interest to you, but not included in the `variants_output.tsv` file, you can retrieve it from the full output file as lined out below.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "discrete-presence",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:41.932607Z",
     "start_time": "2021-03-05T08:49:41.926599Z"
    }
   },
   "outputs": [],
   "source": [
    "full_output_path = tar_ball_out/'release'/'built_final.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "simple-reserve",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:46.059358Z",
     "start_time": "2021-03-05T08:49:41.936792Z"
    }
   },
   "outputs": [],
   "source": [
    "df_full = read_tsv(full_output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "civil-sender",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:46.110393Z",
     "start_time": "2021-03-05T08:49:46.062466Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(40416, 379)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_full.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aware-clear",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:46.140970Z",
     "start_time": "2021-03-05T08:49:46.136063Z"
    }
   },
   "outputs": [],
   "source": [
    "# 2 fields not available in the variants_output.tsv\n",
    "fields_of_interest = ['EUR_Allele_frequency_1000_Genomes', 'AFR_Allele_frequency_1000_Genomes']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dried-glenn",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-04T13:25:41.699844Z",
     "start_time": "2021-03-04T13:25:41.690236Z"
    }
   },
   "source": [
    "We use the internal VCF like representation as directly determined by the source data as join key between the 2 datasets. In the `variants_output.tsv` file the `pyhgvs_Genomic_Coordinate_38` got renamed to `genomic_vcf38_source`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "automatic-change",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:46.747362Z",
     "start_time": "2021-03-05T08:49:46.145337Z"
    }
   },
   "outputs": [],
   "source": [
    "join_key_reduced = 'genomic_vcf38_source'\n",
    "join_key_full = 'pyhgvs_Genomic_Coordinate_38'\n",
    "\n",
    "df_joined = df.merge(df_full[fields_of_interest + [join_key_full]], \n",
    "                 left_on=join_key_reduced, \n",
    "                 right_on=join_key_full,\n",
    "                 how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "alien-possession",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:46.762152Z",
     "start_time": "2021-03-05T08:49:46.750377Z"
    }
   },
   "outputs": [],
   "source": [
    "assert df_full.shape[0] == df_joined.shape[0], \"Not the same amount of variants! Variants lost during joining?\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "welcome-parent",
   "metadata": {},
   "source": [
    "Here we are with the `variants_output.tsv` augmented by some more fields."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "genuine-animation",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:47.073754Z",
     "start_time": "2021-03-05T08:49:46.768113Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomic_hgvs_38</th>\n",
       "      <th>EUR_Allele_frequency_1000_Genomes</th>\n",
       "      <th>AFR_Allele_frequency_1000_Genomes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NC_000013.11:g.32315519C&gt;T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NC_000013.11:g.32315532C&gt;T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_000013.11:g.32315545G&gt;A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>NC_000013.11:g.32315584G&gt;T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>NC_000013.11:g.32315643A&gt;G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40367</th>\n",
       "      <td>NC_000017.11:g.43125374C&gt;A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40368</th>\n",
       "      <td>NC_000017.11:g.43125376G&gt;A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40372</th>\n",
       "      <td>NC_000017.11:g.43125443A&gt;G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40373</th>\n",
       "      <td>NC_000017.11:g.43125449A&gt;G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40376</th>\n",
       "      <td>NC_000017.11:g.43125475T&gt;G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4349 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  genomic_hgvs_38  EUR_Allele_frequency_1000_Genomes  \\\n",
       "41     NC_000013.11:g.32315519C>T                                0.0   \n",
       "46     NC_000013.11:g.32315532C>T                                0.0   \n",
       "54     NC_000013.11:g.32315545G>A                                0.0   \n",
       "58     NC_000013.11:g.32315584G>T                                0.0   \n",
       "65     NC_000013.11:g.32315643A>G                                0.0   \n",
       "...                           ...                                ...   \n",
       "40367  NC_000017.11:g.43125374C>A                                0.0   \n",
       "40368  NC_000017.11:g.43125376G>A                                0.0   \n",
       "40372  NC_000017.11:g.43125443A>G                                0.0   \n",
       "40373  NC_000017.11:g.43125449A>G                                0.0   \n",
       "40376  NC_000017.11:g.43125475T>G                                0.0   \n",
       "\n",
       "       AFR_Allele_frequency_1000_Genomes  \n",
       "41                                0.0023  \n",
       "46                                0.0189  \n",
       "54                                0.0000  \n",
       "58                                0.0015  \n",
       "65                                0.0000  \n",
       "...                                  ...  \n",
       "40367                             0.0000  \n",
       "40368                             0.0000  \n",
       "40372                             0.0159  \n",
       "40373                             0.0000  \n",
       "40376                             0.0000  \n",
       "\n",
       "[4349 rows x 3 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_of_interest_non_na = ~df_joined[fields_of_interest[-1]].isna() # arbitrarily picking the last field of interest\n",
    "cols_for_display = ['genomic_hgvs_38'] + fields_of_interest\n",
    "\n",
    "df_joined[cols_for_display][field_of_interest_non_na]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stunning-overhead",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-02T14:28:58.590914Z",
     "start_time": "2021-03-02T14:28:58.586832Z"
    }
   },
   "source": [
    "### Full File has Unnormalized Coordinates\n",
    "\n",
    "In the `variants_output.tsv` file the `pos`, `ref` and `alt` fields may have different values than `Pos`, `Ref` and `Alt` in the full file, since in the former we use a normalized representation of a variant, in the latter the value as it came from the source.\n",
    "The variants are normalized by right shuffling with respect to the cDNA reference.\n",
    "\n",
    "\n",
    "Variants with a changed representation can be determined as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "identical-burst",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-05T08:49:47.315356Z",
     "start_time": "2021-03-05T08:49:47.124132Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>genomic_vcf38</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>genomic_hgvs_38</th>\n",
       "      <th>cdna</th>\n",
       "      <th>protein</th>\n",
       "      <th>source</th>\n",
       "      <th>clinical_significance_enigma</th>\n",
       "      <th>...</th>\n",
       "      <th>individuals_lovd</th>\n",
       "      <th>literature_source_exlovd</th>\n",
       "      <th>method_clinvar</th>\n",
       "      <th>remarks_lovd</th>\n",
       "      <th>source_url</th>\n",
       "      <th>submitter_clinvar</th>\n",
       "      <th>submitters_lovd</th>\n",
       "      <th>variant_frequency_lovd</th>\n",
       "      <th>url_enigma</th>\n",
       "      <th>flags_gnomad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32314861:AA&gt;A</td>\n",
       "      <td>32314861</td>\n",
       "      <td>AA</td>\n",
       "      <td>A</td>\n",
       "      <td>NC_000013.11:g.32314862del</td>\n",
       "      <td>NM_000059.3:c.-845del</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not in 442 controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>2/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32315357:GCCTGACAAGGAATTTCCTTTCGCCACAC...</td>\n",
       "      <td>32315357</td>\n",
       "      <td>GCCTGACAAGGAATTTCCTTTCGCCACACTGAGAAATACCCGCAGC...</td>\n",
       "      <td>G</td>\n",
       "      <td>NC_000013.11:g.32315358_32316294del</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>research</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV0...</td>\n",
       "      <td>King_Laboratory,_University_of_Washington</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32315426:GG&gt;G</td>\n",
       "      <td>32315426</td>\n",
       "      <td>GG</td>\n",
       "      <td>G</td>\n",
       "      <td>NC_000013.11:g.32315427del</td>\n",
       "      <td>NM_000059.3:c.-280del</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>expression cloning luciferase assay no effect,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>1/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32315429:G&gt;GGTGCGTGTG</td>\n",
       "      <td>32315429</td>\n",
       "      <td>G</td>\n",
       "      <td>GGTGCGTGTG</td>\n",
       "      <td>NC_000013.11:g.32315430_32315438dup</td>\n",
       "      <td>NM_000059.3:c.-277_-269dup</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>LOVD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not in 442 controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lez J Burke (Brisbane,AU)</td>\n",
       "      <td>1/6603 cases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32315523:G&gt;GTGGCACTGCTG</td>\n",
       "      <td>32315523</td>\n",
       "      <td>G</td>\n",
       "      <td>GTGGCACTGCTG</td>\n",
       "      <td>NC_000013.11:g.32315524_32315534dup</td>\n",
       "      <td>NM_000059.3:c.-183_-173dup</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>GnomAD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20731</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32399542:AAATTA&gt;A</td>\n",
       "      <td>32399542</td>\n",
       "      <td>AAATTA</td>\n",
       "      <td>A</td>\n",
       "      <td>NC_000013.11:g.32399543_32399547del</td>\n",
       "      <td>NM_000059.3:c.*773_*777del</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>GnomAD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20737</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32399608:T&gt;TT</td>\n",
       "      <td>32399608</td>\n",
       "      <td>T</td>\n",
       "      <td>TT</td>\n",
       "      <td>NC_000013.11:g.32399609dup</td>\n",
       "      <td>NM_000059.3:c.*839dup</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>GnomAD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20739</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>chr13:g.32399608:TT&gt;T</td>\n",
       "      <td>32399608</td>\n",
       "      <td>TT</td>\n",
       "      <td>T</td>\n",
       "      <td>NC_000013.11:g.32399609del</td>\n",
       "      <td>NM_000059.3:c.*839del</td>\n",
       "      <td>NP_000050.2:p.?</td>\n",
       "      <td>ENIGMA,1000_Genomes,ClinVar,LOVD,GnomAD</td>\n",
       "      <td>Benign</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clinical_testing,curation</td>\n",
       "      <td>Class 1 not pathogenic based on frequency &gt;1% ...</td>\n",
       "      <td>http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV0...</td>\n",
       "      <td>Evidence-based_Network_for_the_Interpretation_...</td>\n",
       "      <td>ENIGMA consortium (Brisbane,AU)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21113</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>chr17:g.43045688:C&gt;CTGGGGGGGG</td>\n",
       "      <td>43045688</td>\n",
       "      <td>C</td>\n",
       "      <td>CTGGGGGGGG</td>\n",
       "      <td>NC_000017.11:g.43045688_43045689insTGGGGGGGG</td>\n",
       "      <td>NM_007294.3:c.5581_5582insCCCCCCCCA</td>\n",
       "      <td>NP_009225.1:p.(His1860_Ser1861insThrProPro)</td>\n",
       "      <td>ESP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23274</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>chr17:g.43051705:GGGCTCACTGCAACCTCTGCCTCCAGGGT...</td>\n",
       "      <td>43051705</td>\n",
       "      <td>GGGCTCACTGCAACCTCTGCCTCCAGGGTTCAAGCGATTCTCCTGC...</td>\n",
       "      <td>G</td>\n",
       "      <td>NC_000017.11:g.43051706_43078153del</td>\n",
       "      <td>NM_007294.3:c.4358-1539_5278-589del</td>\n",
       "      <td>NP_009225.1:p.?</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clinical_testing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV0...</td>\n",
       "      <td>Breast_Cancer_Information_Core_(BIC)_(BRCA1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2507 rows × 220 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_symbol                                      genomic_vcf38  \\\n",
       "7           BRCA2                              chr13:g.32314861:AA>A   \n",
       "20          BRCA2  chr13:g.32315357:GCCTGACAAGGAATTTCCTTTCGCCACAC...   \n",
       "22          BRCA2                              chr13:g.32315426:GG>G   \n",
       "23          BRCA2                      chr13:g.32315429:G>GGTGCGTGTG   \n",
       "39          BRCA2                    chr13:g.32315523:G>GTGGCACTGCTG   \n",
       "...           ...                                                ...   \n",
       "20731       BRCA2                          chr13:g.32399542:AAATTA>A   \n",
       "20737       BRCA2                              chr13:g.32399608:T>TT   \n",
       "20739       BRCA2                              chr13:g.32399608:TT>T   \n",
       "21113       BRCA1                      chr17:g.43045688:C>CTGGGGGGGG   \n",
       "23274       BRCA1  chr17:g.43051705:GGGCTCACTGCAACCTCTGCCTCCAGGGT...   \n",
       "\n",
       "            pos                                                ref  \\\n",
       "7      32314861                                                 AA   \n",
       "20     32315357  GCCTGACAAGGAATTTCCTTTCGCCACACTGAGAAATACCCGCAGC...   \n",
       "22     32315426                                                 GG   \n",
       "23     32315429                                                  G   \n",
       "39     32315523                                                  G   \n",
       "...         ...                                                ...   \n",
       "20731  32399542                                             AAATTA   \n",
       "20737  32399608                                                  T   \n",
       "20739  32399608                                                 TT   \n",
       "21113  43045688                                                  C   \n",
       "23274  43051705  GGGCTCACTGCAACCTCTGCCTCCAGGGTTCAAGCGATTCTCCTGC...   \n",
       "\n",
       "                alt                               genomic_hgvs_38  \\\n",
       "7                 A                    NC_000013.11:g.32314862del   \n",
       "20                G           NC_000013.11:g.32315358_32316294del   \n",
       "22                G                    NC_000013.11:g.32315427del   \n",
       "23       GGTGCGTGTG           NC_000013.11:g.32315430_32315438dup   \n",
       "39     GTGGCACTGCTG           NC_000013.11:g.32315524_32315534dup   \n",
       "...             ...                                           ...   \n",
       "20731             A           NC_000013.11:g.32399543_32399547del   \n",
       "20737            TT                    NC_000013.11:g.32399609dup   \n",
       "20739             T                    NC_000013.11:g.32399609del   \n",
       "21113    CTGGGGGGGG  NC_000017.11:g.43045688_43045689insTGGGGGGGG   \n",
       "23274             G           NC_000017.11:g.43051706_43078153del   \n",
       "\n",
       "                                      cdna  \\\n",
       "7                    NM_000059.3:c.-845del   \n",
       "20                                    None   \n",
       "22                   NM_000059.3:c.-280del   \n",
       "23              NM_000059.3:c.-277_-269dup   \n",
       "39              NM_000059.3:c.-183_-173dup   \n",
       "...                                    ...   \n",
       "20731           NM_000059.3:c.*773_*777del   \n",
       "20737                NM_000059.3:c.*839dup   \n",
       "20739                NM_000059.3:c.*839del   \n",
       "21113  NM_007294.3:c.5581_5582insCCCCCCCCA   \n",
       "23274  NM_007294.3:c.4358-1539_5278-589del   \n",
       "\n",
       "                                           protein  \\\n",
       "7                                  NP_000050.2:p.?   \n",
       "20                                            None   \n",
       "22                                 NP_000050.2:p.?   \n",
       "23                                 NP_000050.2:p.?   \n",
       "39                                 NP_000050.2:p.?   \n",
       "...                                            ...   \n",
       "20731                              NP_000050.2:p.?   \n",
       "20737                              NP_000050.2:p.?   \n",
       "20739                              NP_000050.2:p.?   \n",
       "21113  NP_009225.1:p.(His1860_Ser1861insThrProPro)   \n",
       "23274                              NP_009225.1:p.?   \n",
       "\n",
       "                                        source clinical_significance_enigma  \\\n",
       "7                                         LOVD                          NaN   \n",
       "20                                     ClinVar                          NaN   \n",
       "22                                        LOVD                          NaN   \n",
       "23                                        LOVD                          NaN   \n",
       "39                                      GnomAD                          NaN   \n",
       "...                                        ...                          ...   \n",
       "20731                                   GnomAD                          NaN   \n",
       "20737                                   GnomAD                          NaN   \n",
       "20739  ENIGMA,1000_Genomes,ClinVar,LOVD,GnomAD                       Benign   \n",
       "21113                                      ESP                          NaN   \n",
       "23274                                  ClinVar                          NaN   \n",
       "\n",
       "       ...  individuals_lovd  literature_source_exlovd  \\\n",
       "7      ...               1.0                       NaN   \n",
       "20     ...               NaN                       NaN   \n",
       "22     ...               2.0                       NaN   \n",
       "23     ...               1.0                       NaN   \n",
       "39     ...               NaN                       NaN   \n",
       "...    ...               ...                       ...   \n",
       "20731  ...               NaN                       NaN   \n",
       "20737  ...               NaN                       NaN   \n",
       "20739  ...               1.0                       NaN   \n",
       "21113  ...               NaN                       NaN   \n",
       "23274  ...               NaN                       NaN   \n",
       "\n",
       "                  method_clinvar  \\\n",
       "7                            NaN   \n",
       "20                      research   \n",
       "22                           NaN   \n",
       "23                           NaN   \n",
       "39                           NaN   \n",
       "...                          ...   \n",
       "20731                        NaN   \n",
       "20737                        NaN   \n",
       "20739  clinical_testing,curation   \n",
       "21113                        NaN   \n",
       "23274           clinical_testing   \n",
       "\n",
       "                                            remarks_lovd  \\\n",
       "7                                    not in 442 controls   \n",
       "20                                                   NaN   \n",
       "22     expression cloning luciferase assay no effect,...   \n",
       "23                                   not in 442 controls   \n",
       "39                                                   NaN   \n",
       "...                                                  ...   \n",
       "20731                                                NaN   \n",
       "20737                                                NaN   \n",
       "20739  Class 1 not pathogenic based on frequency >1% ...   \n",
       "21113                                                NaN   \n",
       "23274                                                NaN   \n",
       "\n",
       "                                              source_url  \\\n",
       "7                                                    NaN   \n",
       "20     http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV0...   \n",
       "22                                                   NaN   \n",
       "23                                                   NaN   \n",
       "39                                                   NaN   \n",
       "...                                                  ...   \n",
       "20731                                                NaN   \n",
       "20737                                                NaN   \n",
       "20739  http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV0...   \n",
       "21113                                                NaN   \n",
       "23274  http://www.ncbi.nlm.nih.gov/clinvar/?term=SCV0...   \n",
       "\n",
       "                                       submitter_clinvar  \\\n",
       "7                                                    NaN   \n",
       "20             King_Laboratory,_University_of_Washington   \n",
       "22                                                   NaN   \n",
       "23                                                   NaN   \n",
       "39                                                   NaN   \n",
       "...                                                  ...   \n",
       "20731                                                NaN   \n",
       "20737                                                NaN   \n",
       "20739  Evidence-based_Network_for_the_Interpretation_...   \n",
       "21113                                                NaN   \n",
       "23274       Breast_Cancer_Information_Core_(BIC)_(BRCA1)   \n",
       "\n",
       "                       submitters_lovd  variant_frequency_lovd  url_enigma  \\\n",
       "7            Lez J Burke (Brisbane,AU)            2/6603 cases         NaN   \n",
       "20                                 NaN                     NaN         NaN   \n",
       "22           Lez J Burke (Brisbane,AU)            1/6603 cases         NaN   \n",
       "23           Lez J Burke (Brisbane,AU)            1/6603 cases         NaN   \n",
       "39                                 NaN                     NaN         NaN   \n",
       "...                                ...                     ...         ...   \n",
       "20731                              NaN                     NaN         NaN   \n",
       "20737                              NaN                     NaN         NaN   \n",
       "20739  ENIGMA consortium (Brisbane,AU)                     NaN         NaN   \n",
       "21113                              NaN                     NaN         NaN   \n",
       "23274                              NaN                     NaN         NaN   \n",
       "\n",
       "       flags_gnomad  \n",
       "7               NaN  \n",
       "20              NaN  \n",
       "22              NaN  \n",
       "23              NaN  \n",
       "39              NaN  \n",
       "...             ...  \n",
       "20731           NaN  \n",
       "20737           NaN  \n",
       "20739           NaN  \n",
       "21113           NaN  \n",
       "23274           NaN  \n",
       "\n",
       "[2507 rows x 220 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query('genomic_vcf38 != genomic_vcf38_source')"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
