def existing_variant():
    return {"Variant_frequency_LOVD": "-", "Genomic_Coordinate_hg37": "-",
            "Literature_source_exLOVD": "Easton DF et al., Am J Hum Genet, 81: 873-883, 2007.",
            "ClinVarAccession_ENIGMA": "SCV000244380", "Discordant": "True", "Condition_category_ENIGMA": "Disease",
            "Date_Last_Updated_ClinVar": "2004-02-20,2015-08-10", "Method_ClinVar": "curation,clinical_testing",
            "AFR_Allele_frequency_1000_Genomes": "-", "EUR_Allele_frequency_1000_Genomes": "-",
            "Variant_in_LOVD": True, "Segregation_LR_exLOVD": "1", "Source_URL": "-",
            "Source": "ENIGMA,ClinVar,LOVD,BIC,exLOVD", "Max_Allele_Frequency": "-",
            "Condition_ID_value_ENIGMA": "BREAST-OVARIAN CANCER, FAMILIAL, SUSCEPTIBILITY TO, 1; BROVCA1 (604370)",
            "Protein_Change": "T1685I", "HGVS_Protein": "p.(Thr1685Ile)", "Co_occurrence_LR_exLOVD": "1.1",
            "SAS_Allele_frequency_1000_Genomes": "-", "Variant_in_exLOVD": True,
            "Allele_frequency_1000_Genomes": "-", "Reference_Sequence": "NM_007294.3",
            "Clinical_Significance_ClinVar": "Pathogenic,Uncertain_significance",
            "Functional_analysis_result_LOVD": "predicted deleterious,as variant control,normal RNA splicing,strong functional effect",
            "AMR_Allele_frequency_1000_Genomes": "-", "Variant_in_ESP": False, "Variant_in_BIC": True,
            "Clinical_significance_ENIGMA": "Pathogenic", "Synonyms": "", "Clinical_classification_BIC": "Pending",
            "Gene_Symbol": "BRCA1",
            "Comment_on_clinical_significance_ENIGMA": "IARC class based on posterior probability from multifactorial likelihood analysis, thresholds for class as per Plon et al. 2008 (PMID: 18951446). Class 5 based on posterior probability = 0.999",
            "Sum_family_LR_exLOVD": "125.89",
            "Assertion_method_ENIGMA": "ENIGMA BRCA1/2 Classification Criteria (2015)",
            "Clinical_significance_citations_ENIGMA": "PMID:20513136", "Variant_in_ENIGMA": True,
            "Posterior_probability_exLOVD": "0.9983", "SIFT_VEP": "deleterious(0)", "Germline_or_Somatic_BIC": "G",
            "Missense_analysis_prior_probability_exLOVD": "0.81", "Genomic_Coordinate_hg36": "-",
            "BIC_Identifier": "5173C>T", "Allele_Frequency": "-",
            "Number_of_family_member_carrying_mutation_BIC": "-",
            "Pathogenicity_research": "Pathogenic(ENIGMA); Pathogenic,Uncertain_significance (ClinVar); Pending (BIC)",
            "Allele_Origin_ClinVar": "germline",
            "Submitter_ClinVar": "Breast_Cancer_Information_Core_(BIC)_(BRCA1),Evidence-based_Network_for_the_Interpretation_of_Germline_Mutant_Alleles_(ENIGMA)",
            "Genomic_Coordinate_hg38": "chr17:43067628:G>A", "Patient_nationality_BIC": "-",
            "Literature_citation_BIC": "-", "HGVS_cDNA": "c.5054C>T",
            "Functional_analysis_technique_LOVD": "Transcription activation in yeast (GAL4-fusions),Multifactorial likelihood-ratio model,transcription activation/PS/BA/BS,Multifactorial likelihood model,RNA analysis,evolutionary conservation analysis",
            "Collection_method_ENIGMA": "Curation", "Allele_frequency_ExAC": "-", "Mutation_type_BIC": "M",
            "Combined_prior_probablility_exLOVD": "0.81", "Minor_allele_frequency_ESP": "-",
            "IARC_class_exLOVD": "5 - Definitely pathogenic", "Clinical_importance_BIC": "unknown",
            "Assertion_method_citation_ENIGMA": "http://enigmaconsortium.org/documents/ENIGMA_Rules_2015-03-26.pdf",
            "EAS_Allele_frequency_1000_Genomes": "-", "Variant_haplotype_LOVD": "-",
            "SCV_ClinVar": "SCV000145277,SCV000244380", "Condition_ID_type_ENIGMA": "OMIM",
            "Ethnicity_BIC": "Western-European,Western European,-", "PolyPhen_VEP": "probably_damaging(0-989)",
            "Date_last_evaluated_ENIGMA": "8/10/15", "Pathogenicity_default": "Pathogenic",
            "Variant_in_ExAC": False,
            "URL_ENIGMA": "http://hci-exlovd.hci.utah.edu/variants.php?select_db=BRCA1&action=search_all&search_Variant%2FDNA=c.5054C>T",
            "Variant_in_ClinVar": True, "Allele_origin_ENIGMA": "Germline",
            "Origin_of_variant_LOVD": "In vitro (cloned)", "Variant_in_1000_Genomes": False}


def new_variant():
    v = existing_variant()
    v["Genomic_Coordinate_hg38"] = "chr17:999999:A>G"
    return v
